### DISCUSSION TITLE
Hardware Optolink-Switch

### DISCUSSION THREAD

**Discussion by GolterU** [philippoo66/optolink-splitter#4] [General]
Hallo Phil,
hast Du die Hardware (Stromlaufplan etc.) des optolink-switch irgendwo dokumentiert?

Gruß
Uwe

**Comment by philippoo66** [comment]
Hi Uwe!

die Hardware ist ein Raspi, an den ein Optolink Kopf (IR) angeschossen ist und USB-Serial Adapter mit CP2102 Chip, z.B. so einer

![grafik](https://github.com/philippoo66/optolink-splitter/assets/122479122/e32451f0-d453-4bac-8f8c-c033eff4b36a)

Wo du was am Raspi genau anschliesst (USB oder Raspi UART Pinne) hängt davon ab, welche Adapter du verwendest (ob IR Kopf mit USB oder den CP2102 mit 2 Stk wieder auf USB usw.)

Du brauchst nix löten oder sonstwas!

Grüsse!
Phil

**Reply by philippoo66** [reply]
![grafik](https://github.com/philippoo66/optolink-splitter/assets/122479122/25172c58-bc1e-4b67-bd8c-d57cf3a708b3)

**Comment by GolterU** [comment]
Hallo Phil,
meine Frage zielte darauf ab, wie Du den versprochenen Parallelbetrieb mit dem vitoconnect gelöst hast. Soweit ich das sehe, kommuniziert der ja pausenlos mit der Heizung/Steuerung, weswegen ich es mir schwierig vorstelle, noch eigene Sende-/Empfangsdaten unterzubringen. Es sei denn, der Raspi oben unterbricht einfach den Datenverkehr vitoconnect<->Heizung und sendet/empfängt dann die eigenen Daten.
Ich hatte das im [USB_Duo-Sniffer](https://github.com/openv/openv/wiki/Bauanleitung-USB_Duo-Sniffer) so gelöst, dass der serielle Datenverkehr des vitoconnect zwar ungestört belauscht werden kann. Für eigenen Datenverkehr darf der aber nicht dazwischenfunken.
Deine Beschreibung des [Optolink-Switch im Wiki](https://github.com/philippoo66/optolink-splitter/wiki) lese ich so, dass es mit dem Raspi allein als Hardware nicht getan ist:
> **Hardware**
Der Optolink-Switch wird erstmal zwischen Optolink-Gerät und das Vitoconnect geschaltet.
...
> 

Oder verstehe ich das falsch?
Vielleicht kannst Du mir bisschen auf die Sprünge helfen.
Gruß
Uwe

**Reply by philippoo66** [reply]
der Switch 'besteht' aus mehreren Tasks, die unabhängig die Requests der unterschiedlichen 'Master' empfangen und puffern, und aus einem Loop zur Abarbeitung der Requests (Kommunikation mit dem Optolink Gerät - Heizung, Wärmepumpe, ...) und Rückgabe der entsprechenden Antworten im entsprechenden Format und ggf. an den MQTT Broker publishen. Im Loop-Task findet auch die 'Verwaltung' der Polling Liste statt.

Das Ganze erfordert natürlich die Interpretation der Telegramme, zumindest die Erkennung, wann eins zuende ist. Aber das Vitoconnect benutzt meinen bisherigen Logs zufolge (abgesehen von Initialisierungssequenz) nur VS2 Telegramme (auf VS2 Geräten), und da kann man ab dem 2. bzw. 3. Byte das Ende ersehen. Im full-raw Modus wird das Telegrammende daran ausgemacht, dass eine bestimmte Zeit kein Bit mehr empfangen wurde. Das ist natürlich langsamer, dafür müsste damit auch KW und GWG gehen.

Dann läuft es aktuell so: ein Vitorequest abarbeiten, einen der drei anderen Requests abarbeiten, ein Vitorequest abarbeiten, den nächsten der drei anderen Requests abarbeiten, ... usw. - sequenziell immer im Kreis. Mal schauen, ob die Priorisierung des Vitoconnect wirklich sein muss.

>Deine Beschreibung des [Optolink-Switch im Wiki](https://github.com/philippoo66/optolink-splitter/wiki) lese ich so, dass es mit dem Raspi allein als Hardware nicht getan ist:

klar, wenn du ein Vitoconnect benutzen willt, musst du auch eins haben. Ein Optolink Gerät (Heizung, Wärmepumpe, ...) muss natürlich auch da sein (ein Lenkrad ohne Auto macht auch kein Sinn). Der Optolink/IR Kopf ist gewöhnlich beim Vitoonnect mit bei und dieses CP2102 Böardchen für 3,50€ muss natürlich besorgt werden. Der Optolink-Switch ist der Raspi (oder irgendein anderer Python-kompatibler Rechner mit Seriellen Schnittstellen) mit meinen py Skripts drauf. 

wird es damit klarer?

Grüsse!

**Reply by philippoo66** [reply]
ps: das Mitschreiben der Vitoconnect Telegramme ist natürlich auch eingebaut, und wenn man wollte, könnte man die Daten  und DP-Addr auch an den MQTT Broker schicken. (das Erstellen der csv's für Viessdata ist auch drin).

btw - schön den Creator des Sniffers kennen zu lernen! Es gibt auch noch einen Splitter, aber den hab ich nie zum Laufen bekommen. Es gibt auch optolink2mqtt und vcontrold für tcpip und was weiss ich noch so. aber der hier kann halt alles in einem und parallel, ist einfach in der Anwendung (die SW selbst ist schon etwas komplexer...) und er funktioniert :-)
